matrix.go

```diff
@@ -8,42 +8,64 @@ import (
 	"net/url"
 )
 
+type Params map[string][]string
+type ParamKeys []string
+
 type Matrix struct {
 	Clients   []string
-	Servers   []string
-	Behaviors []string
+	ParamKeys ParamKeys
+	Params    Params
 }
 
 type Results []Result
 
 func BeginMatrixTest(matrix Matrix) Results {
 
-	len := len(matrix.Clients) * len(matrix.Servers) * len(matrix.Behaviors)
-	results := make(Results, 0, len)
+	testCases := []TestCase{}
 
 	for _, client := range matrix.Clients {
-		for _, server := range matrix.Servers {
-			for _, behavior := range matrix.Behaviors {
-
-				testCase := TestCase{
-					Client:   client,
-					Server:   server,
-					Behavior: behavior,
-				}
-
-				result := ExecuteTestCase(testCase)
-				results = append(results, result)
-			}
-		}
+		collect(&testCases, client, matrix, 0, make(map[string]string, 0))
+	}
+
+	// TODO - the testCases end up being the same here, something about the pointer usecase is fucked
+	// suggest just switching to channels at this point...
+
+	results := make(Results, 0)
+	for _, testCase := range testCases {
+
+		fmt.Println("TESTCASE", testCase)
+
+		result := ExecuteTestCase(testCase)
+		results = append(results, result)
 	}
 
 	return results
 }
 
+func collect(testCases *[]TestCase, client string, matrix Matrix, index int, params map[string]string) {
+
+	if index == len(matrix.ParamKeys) {
+		testCase := TestCase{
+			Client: client,
+			Params: params,
+		}
+
+		*testCases = append(*testCases, testCase)
```

Comment:
I'm looking to pass a "results" bag into collect, from where I will append all the testCases to it, since this function is recursive and can't return the testCases directly. @prashantv - am I thinking about this wrong?